File: typescript/src/cli/ui/terminal.ts
/**
 * Asking the terminal what it looks like (2026-09-24).
 *
 * The background colour is the one fact a palette cannot guess: a shaded band
 * that reads as a gentle lift on one dark theme is a grey slab on another, and
 * is wrong on every light one. Gemini CLI and Codex both ask (OSC 11) and mix
 * their surfaces from the answer; so does this.
 *
 * The query is sent together with Primary Device Attributes (DA1), which every
 * terminal answers, and terminals answer in order: DA1 arriving first means
 * the terminal ignores OSC 11, and the wait ends there instead of at a
 * timeout. The wait always ends at DA1, never at the colour, so neither reply
 * can land in the input box later as stray characters. tmux answers DA1 itself and forwards OSC 11 only with
 * passthrough on, so under tmux this usually reports "unknown"; the palette's
 * own values are the fallback.
 */

const OSC11 = /\x1b\]11;rgb:([0-9a-fA-F]{1,4})\/([0-9a-fA-F]{1,4})\/([0-9a-fA-F]{1,4})/;
const DA1 = /\x1b\[\?[0-9;]*c/;

function channel(hex: string): string {
  const value = parseInt(hex, 16) / (16 ** hex.length - 1);
  return Math.round(value * 255)
    .toString(16)
    .padStart(2, '0');
}

/** Parses a terminal's reply; `#rrggbb` for an OSC 11 answer, `null` when there is none (yet). */
export function parseBackgroundReply(reply: string): { background: string | null; complete: boolean } {
  const osc = OSC11.exec(reply);
  const background = osc ? `#${channel(osc[1])}${channel(osc[2])}${channel(osc[3])}` : null;
  // Complete only once DA1 has come: it follows any OSC 11 answer, and
  // stopping at the colour left the DA1 reply to land in the input box.
  return { background, complete: DA1.test(reply) };
}

/** The terminal's background colour as `#rrggbb`, or null when it will not say. */
export function queryBackground(
  input: NodeJS.ReadStream = process.stdin,
  output: NodeJS.WriteStream = process.stdout,
  timeoutMs = 400,
): Promise<string | null> {
  if (!input.isTTY || !output.isTTY || typeof input.setRawMode !== 'function') return Promise.resolve(null);
  return new Promise((resolve) => {
    let reply = '';
    const wasRaw = input.isRaw;
    const finish = (value: string | null) => {
      clearTimeout(timer);
      input.removeListener('data', onData);
      if (!wasRaw) input.setRawMode(false);
      input.pause();
      resolve(value);
    };
    const onData = (data: Buffer | string) => {
      reply += typeof data === 'string' ? data : data.toString('latin1');
      const parsed = parseBackgroundReply(reply);
      if (parsed.complete) finish(parsed.background);
    };
    const timer = setTimeout(() => finish(parseBackgroundReply(reply).background), timeoutMs);
    input.setRawMode(true);
    input.on('data', onData);
    input.resume();
    output.write('\x1b]11;?\x1b\\\x1b[c');
  });
}

/**
 * Where the cursor is (2026-09-25): the input box asks once, as it appears, so
 * it knows whether growing (the `/` menu) scrolled the terminal, and by how
 * much to scroll back when it shrinks (`input.ts`).
 *
 * Sent with DA1 for the same reason as the colour query: the wait ends at DA1.
 * Anything else read meanwhile is typing that arrived early (a person types
 * the next message while the reply is still printing), and it is handed back
 * to the input unchanged rather than lost.
 */
const CPR = /\x1b\[(\d+);(\d+)R/;

/** Parses a reply to `ESC[6n ESC[c`: the 1-based row, whether DA1 came, and the bytes that were not the reply. */
export function parseCursorReply(reply: string): { row: number | null; complete: boolean; rest: string } {
  const cpr = CPR.exec(reply);
  const da1 = DA1.exec(reply);
  let rest = reply;
  if (cpr) rest = rest.replace(cpr[0], '');
  if (da1) rest = rest.replace(da1[0], '');
  return { row: cpr ? Number(cpr[1]) : null, complete: Boolean(da1), rest };
}

/** The cursor's row (1-based), or null when the terminal will not say. */
export function queryCursorRow(
  input: NodeJS.ReadStream = process.stdin,
  output: NodeJS.WriteStream = process.stdout,
  timeoutMs = 250,
): Promise<number | null> {
  if (!input.isTTY || !output.isTTY || typeof input.setRawMode !== 'function') return Promise.resolve(null);
  return new Promise((resolve) => {
    let reply = '';
    const wasRaw = input.isRaw;
    const finish = () => {
      clearTimeout(timer);
      input.removeListener('data', onData);
      if (!wasRaw) input.setRawMode(false);
      input.pause();
      const parsed = parseCursorReply(reply);
      if (parsed.rest) input.unshift(Buffer.from(parsed.rest, 'latin1'));
      resolve(parsed.row);
    };
    const onData = (data: Buffer | string) => {
      reply += typeof data === 'string' ? data : data.toString('latin1');
      if (parseCursorReply(reply).complete) finish();
    };
    const timer = setTimeout(finish, timeoutMs);
    input.setRawMode(true);
    input.on('data', onData);
    input.resume();
    output.write('\x1b[6n\x1b[c');
  });
}

File: python/webagents/cli/ui/terminal.py
"""
Asking the terminal what it looks like, and listening to it while a reply
streams (2026-09-24).

BACKGROUND. The shaded bands behind a sent message and a code block are mixed
from the terminal's own background, which only the terminal knows: a lift
that reads as gentle on one dark theme is a grey slab on another, and wrong on
every light one. The query (OSC 11) goes out together with Primary Device
Attributes (DA1), which every terminal answers, in order: DA1 arriving first
means the terminal ignores OSC 11, so the wait ends there instead of at a
timeout, and no late reply lands in the input box as stray characters. The
TypeScript chat asks the same way (`typescript/src/cli/ui/terminal.ts`).

KEYS WHILE A REPLY STREAMS. Between prompts nothing reads the keyboard, so
the terminal echoed whatever was typed straight into the middle of the answer,
and Esc did nothing. `stream_keys` turns echo off, reads keys itself, and
calls back on Esc; Ctrl+C still arrives as SIGINT, which the chat already
handles. POSIX terminals only; elsewhere both are a no-op.
"""

from __future__ import annotations

import os
import re
import select
import sys
import time
from contextlib import contextmanager
from typing import Callable, Iterator, Optional, Tuple

_OSC11 = re.compile(rb"\x1b\]11;rgb:([0-9a-fA-F]{1,4})/([0-9a-fA-F]{1,4})/([0-9a-fA-F]{1,4})")
_DA1 = re.compile(rb"\x1b\[\?[0-9;]*c")


def _channel(value: bytes) -> str:
    scale = int(value, 16) / (16 ** len(value) - 1)
    return f"{round(scale * 255):02x}"


def parse_background_reply(reply: bytes) -> Tuple[Optional[str], bool]:
    """`(#rrggbb or None, complete)` for what the terminal has said so far.

    Complete only once the DA1 reply has come: terminals answer in order, so
    it follows any OSC 11 answer. Stopping at the colour (2026-09-24) left the
    DA1 reply to arrive after the terminal was put back in cooked mode, and
    the person saw it echoed as `^[[?62;22c` above the chat.
    """
    osc = _OSC11.search(reply)
    background = f"#{_channel(osc.group(1))}{_channel(osc.group(2))}{_channel(osc.group(3))}" if osc else None
    return background, bool(_DA1.search(reply))


def _posix_tty(stream) -> bool:
    try:
        return os.name == "posix" and stream.isatty()
    except (AttributeError, ValueError):
        return False


def query_background(timeout: float = 0.4) -> Optional[str]:
    """The terminal's background colour as `#rrggbb`, or None when it will not say."""
    if not (_posix_tty(sys.stdin) and _posix_tty(sys.stdout)):
        return None
    import termios
    import tty

    fd = sys.stdin.fileno()
    try:
        saved = termios.tcgetattr(fd)
    except termios.error:
        return None
    reply = b""
    try:
        tty.setraw(fd)
        os.write(sys.stdout.fileno(), b"\x1b]11;?\x1b\\\x1b[c")
        deadline = time.monotonic() + timeout
        while True:
            remaining = deadline - time.monotonic()
            if remaining <= 0:
                break
            ready, _, _ = select.select([fd], [], [], remaining)
            if not ready:
                break
            reply += os.read(fd, 1024)
            background, complete = parse_background_reply(reply)
            if complete:
                return background
    except OSError:
        return None
    finally:
        termios.tcsetattr(fd, termios.TCSADRAIN, saved)
    return parse_background_reply(reply)[0]


@contextmanager
def stream_keys(on_escape: Callable[[], None], loop=None) -> Iterator[None]:
    """While a reply streams: no echo, and Esc calls `on_escape`.

    The terminal stays in cbreak mode, so Ctrl+C still raises SIGINT. Other
    keys are read and dropped, so typing ahead does not scribble over the
    answer. An arrow key's escape sequence arrives whole and is not Esc.
    """
    if not _posix_tty(sys.stdin):
        yield
        return
    import asyncio
    import termios
    import tty

    fd = sys.stdin.fileno()
    try:
        saved = termios.tcgetattr(fd)
    except termios.error:
        yield
        return
    loop = loop or asyncio.get_event_loop()

    def _readable() -> None:
        try:
            data = os.read(fd, 1024)
        except OSError:
            return
        if data == b"\x1b":
            on_escape()

    tty.setcbreak(fd)
    loop.add_reader(fd, _readable)
    try:
        yield
    finally:
        loop.remove_reader(fd)
        termios.tcsetattr(fd, termios.TCSADRAIN, saved)
